src/server/pkg/hashtree/hashtree.go

```diff
@@ -0,0 +1,425 @@
+package hashtree
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	pathlib "path"
+
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+)
+
+// updateHash updates the hash of the node N at 'path'. If this changes N's
+// hash, that will render the hash of N's parent (if any) invalid, and this
+// must be called for all parents of 'path' back to the root.
+func (h *HashTree) updateHash(path string) error {
+	n, ok := h.Fs[path]
+	if !ok {
+		return errorf(Internal, "Could not find node \"%s\" to update hash", path)
+	}
+
+	// Compute hash of 'n'
+	var b bytes.Buffer
+	if n.DirNode != nil {
+		// PutFile keeps n.DirNode.Children sorted, so the order is stable
+		for _, child := range n.DirNode.Children {
+			n, ok := h.Fs[join(path, child)]
+			if !ok {
+				return errorf(Internal, "could not find node for \"%s\" while "+
+					"updating hash of \"%s\"", join(path, child), path)
+			}
+			// Write Name and Hash
+			_, err := b.WriteString(fmt.Sprintf("%s:%s:", n.Name, n.Hash))
+			if err != nil {
+				return errorf(Internal, "error updating hash of file at \"%s\": %s",
+					path, err)
+			}
+		}
+	} else if n.FileNode != nil {
+		for _, blockRef := range n.FileNode.BlockRefs {
+			_, err := b.WriteString(fmt.Sprintf("%s:%d:%d:",
+				blockRef.Block.Hash, blockRef.Range.Lower, blockRef.Range.Upper))
+			if err != nil {
+				return errorf(Internal, "error updating hash of dir at \"%s\": %s",
+					path, err)
+			}
+		}
+	} else {
+		return errorf(Internal,
+			"malformed node at \"%s\": it's neither a file nor a directory", path)
+	}
+
+	// Update hash of 'n'
+	cksum := sha256.Sum256(b.Bytes())
+	n.Hash = cksum[:]
+	return nil
+}
+
+// updateFn is used by 'visit'. The first parameter is the node being visited,
+// the second parameter is the path of that node, and the third parameter is the
+// child of that node from the 'path' argument to 'visit'.
+//
+// The *Node argument is guaranteed to have DirNode set (if it's not nil)--visit
+// returns a 'PathConflict' error otherwise.
+type updateFn func(*Node, string, string) error
+
+// This can be passed to visit() to detect PathConflict errors early
+func nop(*Node, string, string) error {
+	return nil
+}
+
+// visit visits every ancestor of 'path' (excluding 'path' itself), leaf to
+// root (i.e.  end of 'path' to beginning), and calls 'update' on each node
+// along the way. For example, if 'visit' is called with 'path'="/path/to/file",
+// then updateFn is called as follows:
+//
+// 1. update(node at "/path/to" or nil, "/path/to", "file")
+// 2. update(node at "/path"    or nil, "/path",    "to")
+// 3. update(node at "/"        or nil, "",         "path")
+//
+// This is useful for propagating changes to size and hash upwards.
+func (h *HashTree) visit(path string, update updateFn) error {
+	for path != "" {
+		parent, child := split(path)
+		pnode, ok := h.Fs[parent]
+		if ok && pnode.DirNode == nil {
+			return errorf(PathConflict, "")
+		}
+		if err := update(pnode, parent, child); err != nil {
```

Comment:
Ah sorry, I thought I replied to this. The answer to both is that this is used when you call PutFile(/path/to/file) and both "/path" and "/path/to" don't exist yet. In that case, the update() function passed by PutFile to visit will create the missing directories.